server/use_cases/add_service_commitments.py

```diff
@@ -39,52 +41,99 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
-        for c in valid_commitments
-    ]
-    if check_time_overlap(valid_shifts):
-        for i in valid_indexes:
-            results[i] = {
+        existing_shifts[c.service_shift_id] for c in valid_commitments]
+    final_valid_commitments = []
+    final_valid_indexes = []
+    for i, idx in enumerate(valid_indexes):
+        current_shift = valid_shifts[i]
+        has_overlap = False
+        # Check for overlap with existing user shifts
+        shifts_to_check = [current_shift]
+        shifts_to_check.extend(existing_user_shifts)
+        if check_time_overlap(shifts_to_check):
+            has_overlap = True
+        if not has_overlap and final_valid_indexes:
+            approved_shifts = [
+                valid_shifts[j] for j in range(i) if j in final_valid_indexes]
+            shifts_to_check = [current_shift]
+            shifts_to_check.extend(approved_shifts)
+            if check_time_overlap(shifts_to_check):
+                has_overlap = True
+        if has_overlap:
+            results[idx] = {
                 "service_commitment_id": None,
-                "success": False
+                "success": False,
+                "message": "Overlapping commitment"
             }
+        else:
+            final_valid_commitments.append(valid_commitments[i])
+            final_valid_indexes.append(idx)
+    if not final_valid_commitments:
+        for i in valid_indexes:
+            if results[i] is None:
+                results[i] = {
+                    "service_commitment_id": None,
+                    "success": False,
+                    "message": "Overlapping commitment"
+                }
         return results
-
-    # insert valid commitments
-    commitments_as_dict = [c.to_dict() for c in valid_commitments]
-    inserted_commitments = (
-        commitments_repo.insert_service_commitments(
-            (commitments_as_dict)
-        )
-    )
-
-    # fill in success results for valid commitments
-    for idx, i in enumerate(valid_indexes):
+    commitments_as_dict = [c.to_dict() for c in final_valid_commitments]
+    inserted_commitments = commitments_repo.insert_service_commitments(
+        commitments_as_dict)
+    for idx, i in enumerate(final_valid_indexes):
         if idx < len(inserted_commitments):
             results[i] = {
-                "service_commitment_id": 
-                    str(inserted_commitments[idx]),
+                "service_commitment_id": str(inserted_commitments[idx]),
                 "success": True
             }
         else:
             results[i] = {
                 "service_commitment_id": None,
-                "success": False
+                "success": False,
+                "message": "Failed to insert commitment"
+            }
+    for i in valid_indexes:
+        if results[i] is None:
+            results[i] = {
+                "service_commitment_id": None,
+                "success": False,
+                "message": "Overlapping commitment"
```

Comment:
This also seems redundant, you already marked the commitments as overlapping, when you discovered that they are overlapping (on line 86)
```suggestion
```